### PR TITLE
[REVIEW] Add Python bindings

### PR DESCRIPTION
**Pull request by shwina**
- [x] Add Cython bindings and Python library
- [x] Update `.gitignore`
- [x] Add docs
- [x] Add black/flake8/isort config
- [ ] Update README

### CODE REVIEW

File: python/nvtx/_lib/lib.pyx
Comment (kkraus14) [inline python/nvtx/_lib/lib.pyx:15]: Should we move these class definitions into a pxd so they can be cimported by someone?

File: python/nvtx/_lib/lib.pyx
Comment (kkraus14) [inline python/nvtx/_lib/lib.pyx:34]: May want to add a comment that a future TODO is to expose a fastpath for ASCII encoding

File: python/nvtx/_lib/lib.pyx
Comment (kkraus14) [inline python/nvtx/_lib/lib.pyx:73]: Should this be a property?
```suggestion
    @property
    def name(self):
        return self._name
```

File: python/nvtx/colors.py
Comment (kkraus14) [inline python/nvtx/colors.py:10]: May want to remove this entry ;)
```suggestion
```

File: python/nvtx/colors.py
Comment (kkraus14) [inline python/nvtx/colors.py:25]: Instead of this check we should just add a dictionary entry for `None` to have a default color.

File: python/nvtx/nvtx.py
Comment (kkraus14) [inline python/nvtx/nvtx.py:89]: If we're keeping the no-op env variable, may want to copy the doc from `annotate` on this so the user can still get a nice docstring even if the env variable turns it to a no-op.

```suggestion
class _annotate_nop:
    def __init__(self, *args, **kwargs):
        pass

    def __enter__(self):
        pass

    def __exit__(self, *args, **kwargs):
        return False

    def __call__(self, func):
        return func


_annotate_nop.__doc__ = annotate.__doc__
_annotate_nop.__init__.__doc__ = annotate.__init__.__doc__
```

File: python/nvtx/__init__.py
Comment (kkraus14) [inline python/nvtx/__init__.py:8]: Do we want to keep this code? I don't think we saw any real change / benefit from it.

File: python/nvtx/tests/test_basic.py
Comment (kkraus14) [inline python/nvtx/tests/test_basic.py:1]: Copyright

File: python/pyproject.toml
Comment (kkraus14) [inline python/pyproject.toml:22]:
```suggestion
requires = ["setuptools", "wheel", "Cython"]

```

File: python/setup.cfg
Comment (kkraus14) [inline python/setup.cfg:20]:
```suggestion
    __init__.py

```

File: python/setup.py
Comment (kkraus14) [inline python/setup.py:1]: Copyright year

File: python/setup.py
Comment (kkraus14) [inline python/setup.py:42]: Python 3.8?

File: python/tools/build-wheels.sh
Comment (kkraus14) [inline python/tools/build-wheels.sh:9]: Are we producing manylinux1 wheels as well?

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: python/tools/build-wheels.sh
Comment (shwina) [reply python/tools/build-wheels.sh:9]: Yeah `auditwheel` sees that we can and automatically produces them: https://test.pypi.org/project/nvtx/0.1.dev1/#files

File: python/nvtx/_lib/lib.pyx
Comment (kkraus14) [inline python/nvtx/_lib/lib.pyx:8]: Is this used anywhere?

File: python/nvtx/__init__.py
Comment (shwina) [reply python/nvtx/__init__.py:1]: Is that:

```
# Copyright 2020  NVIDIA Corporation.  All rights reserved.
#
# Licensed under the Apache License v2.0 with LLVM Exceptions.
# See https://llvm.org/LICENSE.txt for license information.
# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```

Also, does the LLVM Exceptions apply to the Python part of the library?

File: python/nvtx/colors.py
Comment (shwina) [reply python/nvtx/colors.py:10]: This will make @jrhemstad sad :)

Comment (jrhemstad) [reply python/nvtx/colors.py:10]: :cry:

Comment (jrhemstad) [reply python/nvtx/colors.py:10]: Can we somehow extend this list of colors in cuDF to add this?

Comment (kkraus14) [reply python/nvtx/colors.py:10]: > Can we somehow extend this list of colors in cuDF to add this?

Yes we can.

File: python/docs/conf.py
Comment (jrhemstad) [reply python/docs/conf.py:25]: Both C and C++ are v3.

File: python/docs/basic.rst
Comment (jrhemstad) [inline python/docs/basic.rst:53]: There should be more documentation that Domains are really intended to be 1 per library. Lots of people tend to abuse Domains when they should really be using Categories.

Comment (jrhemstad) [reply python/docs/basic.rst:53]: Speaking of which, did Categories not make v1?

Comment (shwina) [reply python/docs/basic.rst:53]: This is `v1` so it's not too late :)

File: python/nvtx/_lib/lib.pyx
Comment (shwina) [reply python/nvtx/_lib/lib.pyx:18]: Done

File: python/nvtx/_lib/lib.pyx
Comment (shwina) [reply python/nvtx/_lib/lib.pyx]: yup - fixed

File: python/nvtx/colors.py
Comment (shwina) [reply python/nvtx/colors.py]: Good idea - done!

File: python/nvtx/nvtx.py
Comment (shwina) [reply python/nvtx/nvtx.py]: Just got rid of the no-op stuff for now

File: python/nvtx/__init__.py
Comment (shwina) [reply python/nvtx/__init__.py]: Nope - deleted

File: python/docs/basic.rst
Comment (shwina) [reply python/docs/basic.rst:53]: Added and documented categories

File: python/nvtx/_lib/lib.pyx
Comment (shwina) [reply python/nvtx/_lib/lib.pyx]: Deleted

File: python/setup.py
Comment (shwina) [reply python/setup.py]: Fixed

File: python/setup.py
Comment (shwina) [reply python/setup.py:47]: Added

File: python/nvtx/tests/test_basic.py
Comment (shwina) [reply python/nvtx/tests/test_basic.py:7]: Done

File: python/docs/basic.rst
Comment (shwina) [reply python/docs/basic.rst:53]: Also a note about when to use domains

File: python/nvtx/__init__.py
Comment (kkraus14) [inline python/nvtx/__init__.py:7]:
```suggestion
```

File: python/nvtx/tests/test_basic.py
Comment (kkraus14) [inline python/nvtx/tests/test_basic.py:9]: I think this pytest dependency needs to be added in a `test_requires` block in `setup.py` and/or `pyproject.toml`

Comment (shwina) [reply python/nvtx/tests/test_basic.py:9]: We have a `test_requirements.txt` (https://github.com/NVIDIA/NVTX/pull/1/files#diff-6afeae61cfaf704a8f0db4151a014574) for this, but we can revisit later if people find it difficult to use that.

**Review by kkraus14 — APPROVED** [review]

**Comment by jcohen-nvidia** [comment]
@shwina I got myself admin permissions finally, and I've created the new branches.  We are going to avoid using "master" as a branch name since LLVM and a number of other projects are starting to switch away from that name.  I made "dev" and "release-v3".  All development work should be direct merges to or pull requests on "dev".  Then when we get to states that are in good shape, we can fast-forward merge all the changes in "dev" to "release-v3" and make a new tag on "release-v3".  So far I've just made a "v3.0.1" tag on that branch.  I set the default branch of the repo to be "release-v3" because that's what people will see when first arriving at the repo page, and that's the thing we want them to get if they just hit Clone or Download.  So now we'll have to remember when creating PRs to not go against the default but to switch it to "dev".  I read various opinions and found a bunch of agreement that this tends to work well in practice, better than having "dev" be default.

So, can you switch this PR to the "dev" branch?

Also, I am going to add you and @kkraus14 to have write access to the repo.  I can merge this into "dev" if you guys have made any needed versioning fixes since the last meeting, where we decided that this would be using the 0.1 or whatever version number, implying that we can make breaking changes without repercussions.

**Comment by shwina** [comment]
Thanks @jcohen-nvidia - I've retargeted to the `dev` branch. As we discussed, the version is set to 0.1. This is good to go from my perspective!

**Review by jrhemstad — COMMENTED** [review]
Should the `python/` dir have it's own README like I do for the cpp dir? 

https://github.com/jrhemstad/NVTX/tree/add-cpp-bindings/cpp

**Comment by shwina** [comment]
> Should the python/ dir have it's own README like I do for the cpp dir?

I'd prefer that we keep a single copy of any Python docs in readthedocs.org, with a link to it in the top-level README. That way we won't have to worry about keeping multiple copies in sync. Once this PR is merged (and once I have write permissions), I'll look into setting up readthedocs, and link to it from the top-level README. Does that sound OK?

**Review by jcohen-nvidia — APPROVED** [review]

**Pull request merged**
